runtime/compiler/env/J9CPU.cpp

```diff
@@ -43,3 +44,29 @@ J9::CPU::getProcessorDescription()
 #endif /* defined(J9VM_OPT_JITSERVER) */
    return _processorDescription;
    }
+
+void
+J9::CPU::applyFeatureMasks(OMRProcessorDesc& processorDescription, const uint32_t* enabledFeatures, size_t size)
+   {
+   OMRPORT_ACCESS_FROM_OMRPORT(TR::Compiler->omrPortLib);
+   memset(_featureMasks.features, 0, OMRPORT_SYSINFO_FEATURES_SIZE*sizeof(uint32_t));
+   for (size_t i = 0; i < size; i++)
+      {
+      omrsysinfo_processor_set_feature(&_featureMasks, enabledFeatures[i], TRUE);
+      }
+   _shouldUseFeatureMasks = true;
+   
+   for (size_t i = 0; i < OMRPORT_SYSINFO_FEATURES_SIZE; i++)
+      {
+      processorDescription.features[i] &= _featureMasks.features[i];
+      }
+   }
+
+bool
+J9::CPU::supportsFeature(uint32_t feature)
+   {
+   OMRPORT_ACCESS_FROM_OMRPORT(TR::Compiler->omrPortLib);
+   if (_shouldUseFeatureMasks)
+      TR_ASSERT_FATAL(TRUE == omrsysinfo_processor_has_feature(&_featureMasks, feature), "new feature usage detected, please enable feature %d in TR::CPU::applyUserOptions()\n", feature);
```

Comment:
Note: if `applyFeatureMasks` can be executed by multiple threads, you cannot expect the features you set previously to remain set in the presence of multiple compilation threads.

Reply:
`applyFeatureMasks` is supposed to be only executed once through `applyUserOptions()` on the `TR::Compiler->target.cpu` singleton. It's a helper function for `applyUserOptions()` and should not be used by the user directly.

I don't like the design either but I have yet to come up with a better design.

Reply:
Maybe I should make `_featureMasks` into a member field for cpu. That way each cpu has their own `_featureMasks` which could make more sense. `applyUserOptions()` is executed once for each newly created cpu object post initialization.

Reply:
actually don't think `_featureMasks` into a member field is a good idea. Also I should separate out the static initialization of `_featureMasks` from `applyUserOptions()`. Initialization of `_featureMasks` should only happen once for the entire JVM while `applyUserOptions()` is performed on each cpu object.